walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,180 @@
+      List<TimeRange> busyByEndTime = orderByEndTime(busyTimeRanges);
+      
+      return findFreeTimes(busyTimeRanges, busyByEndTime, duration);
+  }
+
+  /**
+    * Reorders a List of TimeRange objects by end time
+    *
+    * @param busy the List that is copied and reordered
+    *
+    * @return a copy of busy ordered by end time
+    */
+  private List<TimeRange> orderByEndTime(List<TimeRange> busy) {
+      List<TimeRange> byEndTime = new ArrayList<TimeRange>(busy);
+      Collections.sort(byEndTime, TimeRange.ORDER_BY_END);
+      return byEndTime;
+  }
+
+  /**
+    * Finds the times in a day when mandatory attendees for a meeting are unavailable.
+    *
+    * @param events the events happening that day
+    * @param mandatory the attendees whose attendance is required at another meeting
+    *
+    * @return an ArrayList of TimeRange signifying when mandatory attendees are at another
+    *         event.
+    */
+  private ArrayList<TimeRange> findBusyTimes(Collection<Event> events, List<String> mandatory) {
+      ArrayList<TimeRange> busy = new ArrayList<TimeRange> ();
+
+      for (Event event : events) {          
+          boolean relevantEvent = false;
+          for (String attendee : event.getAttendees()) {
+              if (mandatory.contains(attendee)) {
+                  relevantEvent = true;
+                  break; // only need confirmation that one mandatory attendee is busy
+              }
+          }
+          if (relevantEvent) {
+              busy.add(event.getWhen());
+          }
+      }
+
+      return busy;
+  }
+
+  /**
+    * Finds the times in a day when mandatory attendees for a meeting are available for a given
+    * duration.
+    *
+    * @param busyTimeRanges the times where at least one attendees is unavailable, sorted by start time
+    * @param busybyEndTime the times where at least one attendees is unavailable, sorted by end time
+    * @param duration the length of the new meeting to be added
+    *
+    * @return an ArrayList of TimeRange signifying when mandatory attendees are available for at least
+    *         time duration.
+    */ 
+  private List<TimeRange> findFreeTimes(List<TimeRange> busyTimeRanges, List<TimeRange> busyByEndTime,
+                                        long duration) {
+      
+      List<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+      int startTimeIndex = 0;
+      int endTimeIndex = 0;
+      
+      int numBusyIntervals = busyTimeRanges.size();
+      int lastBusyTime = busyByEndTime.get(numBusyIntervals-1).end();
+
+      // handle beginning of the day before busy intervals begin
+      TimeRange possibleTime = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, 
+                                                      busyTimeRanges.get(endTimeIndex).start(), false);
+      if(possibleTime.duration() >= duration) {
+            meetingTimes.add(possibleTime);
+      }
+      endTimeIndex++;
```

Comment:
Perhaps a silly question.. why do we increment here? Aren't we skipping the very last event then?

Reply:
See previous comment -- endTimeIndex is used as an indicator of the end of a free block of time. After checking the first block of free time at the start of the day, we want to move the endTimeIndex to the front of the second block of busy time.